Mathlib/Topology/VectorBundle/Riemannian.lean

```diff
@@ -0,0 +1,241 @@
+    Continuous (fun b ↦ ⟪v b, w b⟫) := by
+  simp only [continuous_iff_continuousAt] at hv hw ⊢
+  exact fun x ↦ (hv x).inner_bundle (hw x)
+
+end Continuous
+
+namespace Bundle
+
+section Construction
+
+variable
+  {B : Type*} [TopologicalSpace B]
+  {F : Type*} [NormedAddCommGroup F] [NormedSpace ℝ F]
+  {E : B → Type*} [TopologicalSpace (TotalSpace F E)]
+  [∀ b, TopologicalSpace (E b)] [∀ b, AddCommGroup (E b)] [∀ b, Module ℝ (E b)]
+  [∀ b, IsTopologicalAddGroup (E b)] [∀ b, ContinuousConstSMul ℝ (E b)]
+  [FiberBundle F E] [VectorBundle ℝ F E]
+
+open Bornology
+
+variable (E) in
+/-- A family of inner product space structures on the fibers of a fiber bundle, defining the same
+topology as the already existing one. One should more often use `ContinuousRiemannianMetric`
+or `ContMDiffRiemannianMetric` to guarantee continuity or smoothness of the scalar product as a
+function of the base point.
+
+This structure is used through `RiemannianBundle` for typeclass inference, to register the inner
+product space structure on the fibers without creating diamonds. -/
+structure RiemannianMetric where
+  /-- The scalar product along the fibers of the bundle. -/
+  inner (b : B) : E b →L[ℝ] E b →L[ℝ] ℝ
+  symm (b : B) (v w : E b) : inner b v w = inner b w v
+  pos (b : B) (v : E b) (hv : v ≠ 0) : 0 < inner b v v
+  /-- The continuity at `0` is automatic when `E b` is isomorphic to a normed space, but since
+  we are not making this assumption here we have to include it. -/
+  continuousAt (b : B) : ContinuousAt (fun (v : E b) ↦ inner b v v) 0
+  isVonNBounded (b : B) : IsVonNBounded ℝ {v : E b | inner b v v < 1}
+
+/-- `Core structure associated to a family of inner products on the fibers of a fiber bundle. This
+is an auxiliary construction to endow the fibers with an inner product space structure without
+creating diamonds. -/
+@[reducible] noncomputable def RiemannianMetric.toCore (g : RiemannianMetric E) (b : B) :
+    InnerProductSpace.Core ℝ (E b) where
+  inner v w := g.inner b v w
+  conj_inner_symm v w := g.symm b w v
+  re_inner_nonneg v := by
+    rcases eq_or_ne v 0 with rfl | hv
+    · simp
+    · simpa using (g.pos b v hv).le
+  add_left v w x := by simp
+  smul_left c v := by simp
+  definite v h := by contrapose! h; exact (g.pos b v h).ne'
+
+variable (E) in
+/-- Class used to create an inner product structure space on the fibers of a fiber bundle, without
+creating diamonds. Use as follows:
+* `instance : RiemannianBundle E := ⟨g⟩` where `g : RiemannianMetric E` registers the inner product
+space on the fibers;
+* `instance : RiemannianBundle E := ⟨g.toRiemannianMetric⟩` where
+`g : ContinuousRiemannianMetric F E` registers the inner product space on the fibers, and the fact
+that it varies continuously (i.e., a `[IsContinuousRiemannianBundle]` instance).
+* `instance : RiemannianBundle E := ⟨g.toRiemannianMetric⟩` where
+`g : ContMDiffRiemannianMetric IB n F E` registers the inner product space on the fibers, and the
```

Comment:
A not-yet-existing forward reference: that's a promise that you'll make a PR for that soon, right? :tada: (Happy to review that one, I'm fine with this.)